File: _0008.java
package String;

/**
 * 字符串转换整数
 *
 * @author yanxingyu 2512466437@qq.com
 * @version 2022/8/10 10:37
 */
public class _0008 {
    public static int myAtoi(String s) {
        int len=s.length();
        if(len==0) return 0;
        int i=0;
        long sum=0;
        boolean flag=true;
        char c=s.charAt(i);
        while (c==' '){
            i++;
            if(i==len) return 0;
            c=s.charAt(i);
        }
        if(c=='-'){
            i++;
            if(i==len) return 0;
            c=s.charAt(i);
            if(c>'9'||c<'0') return 0;
            flag=false;
        }
        if(c=='+'){
            i++;
            if(i==len) return 0;
            c=s.charAt(i);
            if(c>'9'||c<'0') return 0;
        }
        while (c<='9'&&c>='0'){
            int n=c-'0';
            sum=sum*10+n;
            if(sum>Integer.MAX_VALUE) {
                if(flag) return Integer.MAX_VALUE;
                else return -1-Integer.MAX_VALUE;
            }

            i++;
            if(i==len) break;
            c=s.charAt(i);
        }
        if(flag) return (int)sum;
        else return (int)(0-sum);

    }
    public static void main(String[] args) {
        String s = "1233564457425345";
        System.out.println(myAtoi(s));
    }
}

File: _0008.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
@author: yanxingyu
@email: 2512466437@qq.com
@time: 2022/8/10 21:37
"""
import sys
from this import s


class Solution:
    def myAtoi(self, s: str) -> int:
        # len : int = len(s)
        INT_MAX = 2 ** 31 - 1
        INT_MIN = -2 ** 31
        sLen = len(s)
        if sLen == 0:
            return 0
        i = 0
        Sum: int = 0
        flag = True
        c = s[i]
        while c == ' ':
            i += 1
            if i == sLen:
                return 0
            c = s[i]
        if c == '-':
            i += 1
            flag = False
            if i == sLen:
                return 0
            c = s[i]
            if c > '9' or c < '0':
                return 0
        if c == '+':
            i += 1
            if i == sLen:
                return 0
            c = s[i]
            if c > '9' or c < '0':
                return 0
        while '9' >= c >= '0':
            n: int = ord(c) - ord('0')
            Sum = Sum * 10 + n
            if Sum > INT_MAX:
                if flag:
                    return INT_MAX
                else:
                    return -1 - INT_MAX
            i += 1
            if i == sLen:
                break
            c = s[i]
        if flag:
            return Sum
        else:
            return 0 - Sum


dia = Solution()
s = "  987934636434587"
result = dia.myAtoi(s)
print(result)
